linear_algebra/tensor_product.lean

```diff
@@ -0,0 +1,378 @@
+free_abelian_group.to_add_comm_group (λ (y : M × N), (r • y.1) ⊗ₛ (y.2)) x
+
+@[reducible] def smul (r : R) : M ⊗ N → M ⊗ N :=
+quotient_add_group.lift _ (smul.aux M N r) $ λ x hx,
+begin
+  induction hx with _ hx _ _ ih _ _ _ _ ih1 ih2,
+  { rcases hx with ⟨_, _, _, hx⟩ | ⟨_, _, _, hx⟩ | ⟨_, _, _, hx⟩;
+    rw [hx, smul.aux]; symmetry; simp [free_abelian_group.coe_def];
+    apply quotient.sound,
+    { rw [smul_add],
+      apply group.in_closure.basic, left,
+      exact ⟨_, _, _, by simp; refl⟩ },
+    { apply group.in_closure.basic, right, left,
+      exact ⟨_, _, _, by simp; refl⟩ },
+    { rw [smul_smul, mul_comm r, ← smul_smul],
+      apply group.in_closure.basic, right, right,
+      exact ⟨_, _, _, by simp; refl⟩ } },
+  { refl },
+  { change smul.aux M N r (-_) = 0,
+    rw [is_add_group_hom.neg (smul.aux M N r), ih, neg_zero] },
+  { change smul.aux M N r (_ + _) = 0,
+    rw [is_add_group_hom.add (smul.aux M N r), ih1, ih2, zero_add] }
+end
+
+def smul.is_add_group_hom (r : R) : is_add_group_hom (smul M N r) :=
+by apply_instance
+local attribute [instance] smul.is_add_group_hom
+
+protected theorem smul_add (r : R) (x y : M ⊗ N) :
+  smul M N r (x + y) = smul M N r x + smul M N r y :=
+is_add_group_hom.add _ _ _
+
+instance : module R (M ⊗ N) :=
+{ smul := smul M N,
+  smul_add := tensor_product.smul_add M N,
+  add_smul := begin
+      intros r s x,
+      apply quotient_add_group.induction_on x,
+      intro z,
+      symmetry,
+      refine @free_abelian_group.to_add_comm_group.unique _ _ _ _ _ ⟨λ p q, _⟩ _ z,
+      { simp [smul, smul.aux] },
+      intro x,
+      cases x with m n,
+      exact eq.symm (quotient.sound $ group.in_closure.basic $ or.inl
+      ⟨_, _, _, by simp [add_smul]; refl⟩),
+    end,
+  mul_smul := begin
+      intros r s x,
+      apply quotient_add_group.induction_on' x,
+      intro z,
+      symmetry,
+      refine @free_abelian_group.to_add_comm_group.unique _ _ _ _ _
+        ⟨λ p q, _⟩ _ z,
+      { simp [tensor_product.smul_add] },
+      intro x,
+      cases x with m n,
+      simp [smul, smul.aux, mul_smul, of]
+    end,
+  one_smul := λ x, quotient.induction_on x $ λ _,
+    eq.symm $ free_abelian_group.to_add_comm_group.unique _ _ $ λ ⟨p, q⟩,
+    by rw [one_smul]; refl,
+  .. tensor_product.add_comm_group M N }
+
+theorem bilinear : is_bilinear_map (@of R _ M N _ _) :=
+{ add_pair := λ m1 m2 n, eq.symm $ sub_eq_zero.1 $ eq.symm $
+    quotient.sound $ group.in_closure.basic $ or.inl
+      ⟨_, _, _, by simp; refl⟩,
+  pair_add := λ m1 m2 n, eq.symm $ sub_eq_zero.1 $ eq.symm $
+    quotient.sound $ group.in_closure.basic $ or.inr $ or.inl
+      ⟨_, _, _, by simp; refl⟩,
+  smul_pair := λ r x y, rfl,
+  pair_smul := λ r x y, eq.symm $ sub_eq_zero.1 $ eq.symm $
+    quotient.sound $ group.in_closure.basic $ or.inr $ or.inr
+      ⟨_, _, _, by simp; refl⟩ }
+
+@[simp] lemma add_of (m₁ m₂ : M) (n : N) :
+  (m₁ + m₂) ⊗ₛ n = m₁ ⊗ₛ n + m₂ ⊗ₛ n :=
+(bilinear M N).add_pair m₁ m₂ n
+

```

Comment:
is there a case for calling this `left_bilinear` or `right_bilinear` or something?

Reply:
I would indeed prefer a name closer to `add_mul` here. I think we need to coin a "short name" for this tensor product operation to be available in this context. Why is there a subscript `s`? It could be called `omul` using symbolic naming, or `tmul` for tensor multiplication, or `muls` since there is an `s`, although that will possibly get confused with `smul`.

Reply:
I will go with `tmul` and `⊗ₜ`